File: apps/web/lib/assistant_section_stream.ts
import { sectionSlug, type AnswerSection } from '@/lib/assistant_sections';

type SectionEntry = {
  id: string;
  title: string;
  text: string;
  finalized: boolean;
};

export type StreamSection = {
  id: string;
  title: string;
  text: string;
  finalized: boolean;
};

/** P4：消费 section_* SSE，按节累积 Markdown，避免流式结构跳变。 */
export class SectionStreamAccumulator {
  private entries = new Map<string, SectionEntry>();
  private order: string[] = [];
  private _active = false;

  get active(): boolean {
    return this._active;
  }

  seedFromPlan(titles: string[] | undefined): void {
    if (!titles?.length) return;
    for (const title of titles) {
      const id = sectionSlug(title);
      if (this.entries.has(id)) continue;
      this.entries.set(id, { id, title, text: '', finalized: false });
      this.order.push(id);
    }
  }

  onStart(payload: { id: string; title: string }): void {
    this._active = true;
    const id = payload.id?.trim() || sectionSlug(payload.title);
    const title = payload.title?.trim() || id;
    if (!this.entries.has(id)) {
      this.entries.set(id, { id, title, text: '', finalized: false });
      this.order.push(id);
      return;
    }
    const existing = this.entries.get(id)!;
    if (title && existing.title !== title) {
      existing.title = title;
    }
  }

  onDelta(payload: { id: string; text: string }): void {
    if (!payload.text) return;
    this._active = true;
    const id = payload.id?.trim();
    if (!id) return;
    let entry = this.entries.get(id);
    if (!entry) {
      entry = { id, title: id, text: '', finalized: false };
      this.entries.set(id, entry);
      this.order.push(id);
    }
    entry.text += payload.text;
  }

  onDone(payload: { id: string; title: string; text: string }): void {
    this._active = true;
    const id = payload.id?.trim() || sectionSlug(payload.title);
    const title = payload.title?.trim() || id;
    let entry = this.entries.get(id);
    if (!entry) {
      entry = { id, title, text: '', finalized: false };
      this.entries.set(id, entry);
      this.order.push(id);
    }
    entry.title = title;
    entry.text = payload.text ?? entry.text;
    entry.finalized = true;
  }

  toMarkdown(): string {
    const parts: string[] = [];
    for (const id of this.order) {
      const entry = this.entries.get(id);
      if (!entry) continue;
      if (!entry.text.trim() && entry.finalized) continue;
      parts.push(`### ${entry.title}\n${entry.text}`.trimEnd());
    }
    return parts.join('\n\n').trim();
  }

  getSections(): AnswerSection[] {
    return this.order
      .map((id) => this.entries.get(id))
      .filter((e): e is SectionEntry => Boolean(e))
      .map((e) => ({ id: e.id, title: e.title }));
  }

  getWrittenSectionIds(): Set<string> {
    const out = new Set<string>();
    for (const id of this.order) {
      const entry = this.entries.get(id);
      if (entry?.text.trim()) out.add(id);
    }
    return out;
  }

  /** 流式增量渲染：按 section 返回正文，已 finalize 的节不再重算。 */
  getRenderableSections(): Array<{
    id: string;
    title: string;
    text: string;
    finalized: boolean;
  }> {
    return this.order
      .map((id) => this.entries.get(id))
      .filter((e): e is SectionEntry => Boolean(e))
      .filter((e) => e.text.trim() || !e.finalized)
      .map((e) => ({
        id: e.id,
        title: e.title,
        text: e.text,
        finalized: e.finalized,
      }));
  }
}

File: apps/mobile/lib/features/assistant/assistant_section_stream.dart
/// P4：消费 section_* SSE，按节累积 Markdown。
library;

import 'assistant_sections.dart';

class StreamSection {
  const StreamSection({
    required this.id,
    required this.title,
    required this.text,
    this.finalized = false,
  });

  final String id;
  final String title;
  final String text;
  final bool finalized;
}

class _SectionEntry {
  _SectionEntry({
    required this.id,
    required this.title,
    this.text = '',
    this.finalized = false,
  });

  final String id;
  String title;
  String text;
  bool finalized;
}

class SectionStreamAccumulator {
  final Map<String, _SectionEntry> _entries = {};
  final List<String> _order = [];
  bool _active = false;

  bool get active => _active;

  void seedFromPlan(List<String>? titles) {
    if (titles == null || titles.isEmpty) return;
    for (final title in titles) {
      final id = sectionSlug(title);
      if (_entries.containsKey(id)) continue;
      _entries[id] = _SectionEntry(id: id, title: title);
      _order.add(id);
    }
    if (_order.isNotEmpty) _active = true;
  }

  void onStart({required String id, required String title}) {
    _active = true;
    final sid = id.trim().isNotEmpty ? id.trim() : sectionSlug(title);
    final stitle = title.trim().isNotEmpty ? title.trim() : sid;
    final existing = _entries[sid];
    if (existing == null) {
      _entries[sid] = _SectionEntry(id: sid, title: stitle);
      _order.add(sid);
      return;
    }
    if (stitle.isNotEmpty) existing.title = stitle;
  }

  void onDelta({required String id, required String text}) {
    if (text.isEmpty) return;
    _active = true;
    final sid = id.trim();
    if (sid.isEmpty) return;
    var entry = _entries[sid];
    if (entry == null) {
      entry = _SectionEntry(id: sid, title: sid);
      _entries[sid] = entry;
      _order.add(sid);
    }
    entry.text += text;
  }

  void onDone({required String id, required String title, required String text}) {
    _active = true;
    final sid = id.trim().isNotEmpty ? id.trim() : sectionSlug(title);
    final stitle = title.trim().isNotEmpty ? title.trim() : sid;
    var entry = _entries[sid];
    if (entry == null) {
      entry = _SectionEntry(id: sid, title: stitle);
      _entries[sid] = entry;
      _order.add(sid);
    }
    entry.title = stitle;
    final incoming = text.trim();
    if (incoming.isNotEmpty &&
        (entry.text.trim().isEmpty ||
            incoming.length >= entry.text.trim().length)) {
      entry.text = incoming;
    }
    entry.finalized = true;
  }

  String toMarkdown() {
    final parts = <String>[];
    for (final sid in _order) {
      final entry = _entries[sid];
      if (entry == null) continue;
      if (entry.text.trim().isEmpty) continue;
      parts.add('### ${entry.title}\n${entry.text}'.trimRight());
    }
    return parts.join('\n\n').trim();
  }

  List<AnswerSection> getSections() {
    return _order
        .map((sid) => _entries[sid])
        .whereType<_SectionEntry>()
        .map((e) => AnswerSection(id: e.id, title: e.title))
        .toList();
  }

  Set<String> getWrittenSectionIds() {
    final out = <String>{};
    for (final sid in _order) {
      final entry = _entries[sid];
      if (entry != null && entry.text.trim().isNotEmpty) {
        out.add(sid);
      }
    }
    return out;
  }

  List<StreamSection> getRenderableSections() {
    return _order
        .map((sid) => _entries[sid])
        .whereType<_SectionEntry>()
        .where((e) => e.text.trim().isNotEmpty || !e.finalized)
        .map(
          (e) => StreamSection(
            id: e.id,
            title: e.title,
            text: e.text,
            finalized: e.finalized,
          ),
        )
        .toList();
  }
}
